iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java

```diff
@@ -123,6 +123,11 @@ public class IotHubTransport implements IotHubListener
     // Used to store the number of milliseconds since epoch that this packet was created for a correlationId
     private final Map<String, Long> correlationStartTimeMillis = new ConcurrentHashMap<>();
 
+    // A job that runs periodically to remove any stale correlation callbacks
+    private final Thread correlationCallbackCleanupThread = new Thread(() -> checkForOldMessages());
+    private static final int CORRELATION_CALLBACK_CLEANUP_PERIOD_MILLISECONDS = 60 * 60 * 1000;
```

Comment:
The only concern I have is that, the original OutOfMemory issue could happen frequently. According to the original [report](https://github.com/Azure/azure-iot-sdk-java/issues/1629), it was reproducible within 15 min. Certainly this depends on the size of environment, but maybe we can decrease the sleeping time length here down to 5 mins or 10 mins perhaps?

Reply:
I'd expect that these callback maps would get cleaned up as each twin operation finished successfully. This cleanup job is moreso here to catch any entries who belonged to twin operations that didn't finish (service never sent the twin response). With that in mind, having this cleanup job run less frequently should be fine since user's shouldn't expect a lot of twin failures.

That said, I'll run some tests locally to make sure that these maps are being cleaned up even without the cleanup job and let you know.